### PR TITLE
[docs-only] draft an adr for extension templates

### PR DESCRIPTION
**Pull request by C0rby**
This PR introduces an ADR for the extension template.
I also want to use this PR to this discuss this topic.

### CODE REVIEW

**Comment by C0rby** [comment]
After discussing this with @butonic and @wkloucek we came to the conclusion that we might need different templates for different audiences.

We internally might want to have the same structure in all of our extensions and therefore could use a template, be it using boilr or copying ocis-hello and replacing stuff.

On the other hand someone who wants to develop a simple extension like a todo app doesn't need everything we have in the boilr template. They just need to know how they can register their extension in the registry/the proxy and how to communicate with the storage layer. But they can use what ever they want to implement configuration, flag parsing, http endpoints and so on.
They might or might not implement metrics or tracing but even if they do the project structure doesn't matter.

We currently also lack some features to simplify the extension model. For example the proxy should make use of the registry to look up the addresses of the services.
Also dynamically registering routes in the proxy and the web config would be nice.

**tl;dr**
Technically the extensions don't need to be in a certain structure and don't need to use the same libraries. But if the extensions should be included into the oCIS binary they should have the same structure and libraries.

We need to document the minimal steps necessary to register an extension.

**Comment by wkloucek** [comment]
@C0rby should we also add the "one-file" extension idea to show off what minimum extensions can be achieved?

**Comment by wkloucek** [comment]
there will be a big change to the "Develop - Extensions" part of our docs in https://github.com/owncloud/ocis/pull/2182

**Comment by C0rby** [comment]
I think we also need to get a common understanding how we want to ship extensions.
In a discussion yesterday I heard that we generally want to compile the extension into a single oCIS binary. With that requirement it would be necessary to have some of the `go-micro` stuff in an extension.

**Comment by C0rby** [comment]
> @C0rby should we also add the "one-file" extension idea to show off what minimum extensions can be achieved?

We can but I think it would first make sense to talk about the expectations for extensions like: https://github.com/owncloud/ocis/pull/2168#issuecomment-863022261. Because with that requirement the simple one file extension wouldn't make sense.

But maybe not every extension will be compiled into the single binary, then the simple file would make sense again.

**Comment by butonic** [comment]
is in draft state.

**Review by micbar — APPROVED** [review]

**Review by butonic — APPROVED** [review]
